Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp

```diff
@@ -0,0 +1,258 @@
+    ~SpielSpeechWrapper();
+
+    Vector<RefPtr<PlatformSpeechSynthesisVoice>> initializeVoiceList();
+    void pause();
+    void resume();
+    void speakUtterance(RefPtr<PlatformSpeechSynthesisUtterance>&&);
+    void cancel();
+    void resetState();
+
+private:
+    RefPtr<PlatformSpeechSynthesisUtterance> m_utterance;
+    const PlatformSpeechSynthesizer& m_platformSynthesizer;
+    GRefPtr<SpielSpeaker> m_speaker;
+    Vector<GRefPtr<SpielVoice>> m_voices;
+};
+
+SpielSpeechWrapper::SpielSpeechWrapper(const PlatformSpeechSynthesizer& synthesizer)
+    : m_platformSynthesizer(synthesizer)
+{
+    ensureGStreamerInitialized();
+    registerWebKitGStreamerElements();
+
+    GRefPtr<GstElement> audioSink = createPlatformAudioSink("speech"_s);
+    if (!audioSink) {
+        GST_ERROR("Failed to create GStreamer audio sink element");
+        return;
+    }
+
+    // Async API can't really be used here because the voices initialization
+    // (PlatformSpeechSynthesizer::initializeVoiceList()) is synchronous.
+    GUniqueOutPtr<GError> error;
+    m_speaker = adoptGRef(spiel_speaker_new_sync(nullptr, &error.outPtr()));
+    if (error) {
+        WTFLogAlways("Spiel speaker failed to initialize: %s", error->message);
+        return;
+    }
+
+    g_object_set(m_speaker.get(), "sink", audioSink.get(), nullptr);
+
+    // TODO: Plumb support for boundaryEventOccurred? Using range-started signal?
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-started", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().didStartSpeaking(*self->m_utterance);
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-finished", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().didFinishSpeaking(*self->m_utterance);
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-canceled", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().didFinishSpeaking(*self->m_utterance);
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-error", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().speakingErrorOccurred(*self->m_utterance);
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "notify::paused", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielSpeaker* speaker) {
+        gboolean isPaused;
+        g_object_get(speaker, "paused", &isPaused, nullptr);
+        if (isPaused)
+            self->m_platformSynthesizer.client().didPauseSpeaking(*self->m_utterance);
+        else
+            self->m_platformSynthesizer.client().didResumeSpeaking(*self->m_utterance);
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "notify::voices", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielSpeaker*) {
+        self->m_platformSynthesizer.client().voicesDidChange();
```

Comment:
Maybe this happens somewhere else, but does the new voice list get reflected to content?

Reply:
I'm not sure what you're asking :) `voicesDidChange()` triggers a `voiceschangedEvent`  (see `SpeechSynthesis::voicesDidChange()`) and a re-initialization of the voices list the next time they are queried from JS.